### ISSUE TITLE
Tor integration (and other Privacy-enhancing technologies)

### ISSUE THREAD

**Issue by eragmus** [voisine/breadwallet#178]
Hello,

Can we integrate Tor into breadwallet?

This is a fairly obvious request and one that I feel must have been suggested before. But, upon a cursory glance, I do not see this officially suggested nor debated!

SPV is great, but the downside is that random nodes from the Bitcoin network are contacted to propagate payments. These nodes may be benign or malicious, SPV does not know (and until the latest update of breadwallet addressing tarpit nodes, it did not distinguish good from bad). Besides stalling transactions, nodes may surveil the network and attempt to build profiles via IP addresses, among other information. To hinder that surveillance, increase privacy of breadwallet users, and thereby increase the integrity of the Bitcoin ecosystem, I think it is high priority to integrate Tor (or something else as effective, but Tor seems by and large the most recommended and vetted) in a seamless, secure, and automatic manner.

Adding further in the vein of privacy, there are other technologies (developed, or in development) that look great, in combination with Tor:
- Stealth addresses (like darkwallet)
- CoinJoin (like darkwallet)
- Threshold Signatures (https://freedom-to-tinker.com/blog/stevenag/threshold-signatures-for-bitcoin-wallets-are-finally-here/)
- Merge Avoidance
- CoinShuffle
- techniques/technologies listed here: https://wiki.unsystem.net/en/index.php/DarkWallet/Best_practices
- etc etc, too many more up-and-coming technologies to mention (like the aforementioned threshold signatures, which was suddenly released into the wild out of the blue!)

Privacy (and Security) must be paramount, and a solid foundation must be built. It's counterproductive to propagate Bitcoin to the masses, and then deal with negative repercussions and bad PR down the road.

**Comment by eragmus** [comment]
Bump...

As an aside, I still find it really unattractive that any transaction sent through breadwallet is easily traceable by the recipient, who can use a block explorer to see the balance of the address that sent the bitcoins. In addition, fungibility is essential for currency and only assured if the history of each bitcoin is unknowable.

Hence, this is why the technologies mentioned in this thread are so key.

**Comment by voisine** [comment]
I agree. Unfortunately all the privacy techniques other than merge avoidance have significant usability issues in their current form, so we will focus on merge avoidance first while continuing to think about how to improve on the others.

**Comment by eragmus** [comment]
Hello, back to privacy again.

**re: CoinJoin**

Darkwallet has their implementation here:
> https://github.com/darkwallet/coinjoin.js

Since the code is already there and has been tested, can we adopt this into breadwallet and have it activate by default, if/when it's feasible?

**re: Stealth addresses**

Similarly, Darkwallet has their implementation here:
> https://github.com/darkwallet/stealth.js

Can we adopt it into breadwallet?

Thanks.

**Comment by voisine** [comment]
Current stealth addresses don't work well with SPV and coinjoin has practical usability issues. Tor is still something we're interested in. Also Jonathan Hope has published some research into efficient merge avoidance techniques. We met with him in Tokyo and he indicated he was interested in working on an implementation for breadwallet, so keep an eye out for work on that front.

**Comment by dabura667** [comment]
> Since the code is already there and has been tested
js != Obj C... so it's not as easy as you make it sound.
Also, darkwallet is not "well tested." It is more like "abandoned" and "the kind of people running into problems are likely to just take the loss and not report any issues, as they are afraid of identifying themselves online."

> Can we adopt it into breadwallet?
No, darkwallet stealth requires holding the full blockchain. Since breadwallet is an SPV wallet that contacts the Bitcoin network directly, this can not be done without creating a centralized server structure like darkwallet's obelisk nodes. breadwallet does not want to create some sort of special server structure like electrum and darkwallet... so no.

> I also want to bump Tor, specifically the iOS implementation.
breadwallet has no way to choose nodes, and will choose them at random. Since there is no peer discovery method to find hidden service nodes, there would be no way to find the other nodes.
And if you are talking about accessing normal nodes through an exit node, you need to read the recent paper about deanonymizing Tor. There would be little to no benefit, and would require bad changes to breadwallet's network processes in order to get it to work. So no.


I was one of the people working on the stealth address implementation with the darkwallet guys.
It has since been abandoned, so I recommend no more wallets support it.

**Comment by eragmus** [comment]
Hi dabura667,

Okay to most of your comments, thanks for clarifying.

Except:
> And if you are talking about accessing normal nodes through an exit node, you need to read the recent paper about deanonymizing Tor. There would be little to no benefit, and would require bad changes to breadwallet's network processes in order to get it to work. So no.

Yes, I was talking about breadwallet having built-in Tor to access bitcoin network nodes. Regarding the 'recent paper about deanonymizing Tor', Steve Dekorte seems to believe that it is not an issue.

* https://twitter.com/voluntarynet/status/637164439649677312

* https://github.com/Voluntarynet/Bitmarkets/issues/29#issuecomment-135891950

"AFAICS, that’s only an issue with how Tor onion hidden services set up routing and shouldn’t be a problem for how we use Tor to talk to the Bitcoin and Bitmessage networks."


re: Stealth

> I was one of the people working on the stealth address implementation with the darkwallet guys. It has since been abandoned in favor for a different implementation which has yet to be decided upon, so I recommend no more wallets support it.

So do you recommend ArcBit (https://github.com/arcbit/arcbit-ios) also drop their implementation (referred to as "Forward/reusable/stealth address support")?

Thanks.

**Comment by eragmus** [comment]
Hi Aaron,

> Current stealth addresses don't work well with SPV and coinjoin has practical usability issues. Tor is still something we're interested in. Also Jonathan Hope has published some research into efficient merge avoidance techniques. We met with him in Tokyo and he indicated he was interested in working on an implementation for breadwallet, so keep an eye out for work on that front.

Unfortunate about stealth + CoinJoin. Good to hear about the merge avoidance, though!

**Comment by dabura667** [comment]
> So do you recommend ArcBit (https://github.com/arcbit/arcbit-ios) also drop their implementation (referred to as "Forward/reusable/stealth address support")?

Nah, tbh I think the stealth address in it's current form is so clunky it's nearly impossible for it to "go mainstream." so I see no harm in 1 or 2 developers trying to somehow make it work for their app.

https://github.com/bitcoin/bips/pull/159
BIP 47 is looking like it will be a better solution.

**Comment by eragmus** [comment]
Okay, thanks dabura667.

What about these other suggestions?

1. BIP 69: Lexicographical Indexing of Inputs and Outputs: position the inputs/outputs of each transaction deterministically after a send, decreasing the effectiveness of block chain snooping (implemented in Electrum + ArcBit + Samourai)

2. 5-8 digit PIN to improve security (Samourai)

3. PIN screen randomized on every load to thwart screen recording spyware and people looking over your shoulder to guess which numbers you tapped (http://samouraiwallet.com/static/public/images/pin-screenshot.png -- Samourai)

4. Randomize the number of change outputs for each send, making gathering wallet address metadata more difficult (Samourai)

5. Avoid linking previous change outputs: parse all previous transactions before building a new transaction to make sure it isn't leaking valuable metadata about your and your wallet (Samourai)

**Comment by voisine** [comment]
1) we order inputs by age and outputs are randomized. This achieves the same goal as bip69

2) longer passcodes are planned, however the real protection isn't in the app passcode, it's in the main phone passcode. 

3) If you have malware that can read the screen, you're pretty much screwed no matter what. The device is compromised.

4-5) this will be part of merge avoidance

**Comment by eragmus** [comment]
Nice!

re: 3) If you have malware that can read the screen, you're pretty much screwed no matter what. The device is compromised.

That's true. What about the situation where someone looks over your shoulder and happens to notice the area of the screen tapped?

People often choose poor passcodes, such as "8-0-8-0", simply because it is easy to tap quickly. If one saw that repetitive movement of tapping, one would be able to do a few tests and perhaps chance upon the correct passcode. (But agreed that the phone passcode and/or touch ID adds some help to reduce this chance).


Another idea:

For those with touch ID, with a send transaction, is it possible to require PIN (it's currently required, and a great idea to require only PIN rather than only touch ID) AND touch ID?

That would further drastically reduce the chance of someone trying to secretly guess the passcode without being in the presence of the owner.

**Comment by voisine** [comment]
Touch ID does not work 100% of the time. It can only be an option to skip the passcode when it works. Having the same button pattern also lets people enter the code more quickly so less likely for shoulder surfers to catch it.

**Comment by eragmus** [comment]
I guess I can't argue with either of those reasons :).

**Comment by eragmus** [comment]
So, regarding CoinJoin's usability issues, please see Chris Belcher's JoinMarket:
> https://github.com/chris-belcher/joinmarket

This project uses the command line (no GUI yet), but it already works and successful coinjoins have been completed. They are now working on creating a plugin for Electrum. Can breadwallet collaborate with JoinMarket to integrate it into the wallet by default?

As I see it, the basic problem with CoinJoin (the usability issue) is that it requires others to also be making transactions at the same time and with a similar or same amount. I'm not sure how popular breadwallet is (how many transactions are being made at any given time using breadwallet), but perhaps it is enough to enable CoinJoin of txs (of the complete amount, or maybe at least a partial amount if that's possible).

**Comment by voisine** [comment]
It would be a really big departure from the current instant mobile app experience of opening the app, authorizing a transaction, and closing it again. With coinjoin the process becomes asyncronous where you often need to wait for a later notification that the tx is ready for signing. For this reason it's pretty far down on the list of ways to improve financial privacy.

**Comment by eragmus** [comment]
Ah, yeah I can see what you mean, as that would adversely affect the user experience. Hypothetically, if in the future TPS over breadwallet was to increase by 10x or even 100x, would that then make CoinJoin / JoinMarket more viable (by shortening the time required to wait in that asynchronous process)?

**Comment by voisine** [comment]
actually reading more about join market, it looks like an interesting way fix the problem, basically there are people in the market sitting there with open offers to do joins for a fee, so a wallet user can just accept an available offer at the time they want to perform a tx. That solves the usability issue, so once this market is up and running and has a reasonably deep liquid pool of join offers at very low competitive prices, then we might consider spending engineering resources implementing.

**Comment by eragmus** [comment]
Yes!!! Haha I was just in the process of editing my post to add that distinction, as I too forgot it.

Looking at the order book, there's a user offering max 10 bitcoins to CoinJoin with, and the total fee is: .0013/100*10+0.00002 = 0.00015 bitcoins, or 150 bits to mix 10 bitcoins. Sounds amazing.

Here are two order book implementations, for reference:

* http://www.joinmarket.io

* http://bitcoin-p2pool.com/joinmarket_orderbook.html

I also want to share the Reddit thread on this topic by the creator, u/belcher_:
> https://www.reddit.com/r/Bitcoin/comments/2zc5tc/joinmarket_increase_the_privacy_of_bitcoin_and/

And some choice quotes, for reference:
> The idea behind JoinMarket is that holders of bitcoin will allow their coins to be mixed with in return for a fee. The mixing will happen in coinjoin transactions.

> Because of the fee paid, coinjoin mixing in JoinMarket will become like an almost-riskless savings account for owners of bitcoins. Their private keys will never leave their computers and they will not sign transactions not to their liking.

> As the risk is very low, the reward will also be low because of competition between fee-earners. It means that you will be able to do a coinjoin very cheaply. We already see that holders of bitcoin are willing to earn just 0.006% per day by lending on the bitfinex exchange, and that contains a substantial risk that bitfinex will go disappear.

> I think JoinMarket will succeed where others have failed because of the practical and realistic incentive structure. People who desire privacy will be able to immediately mix their coins with a huge number of other people's coins, in return for a low fee. People who own bitcoins will have access to a safe investment.

**Comment by voisine** [comment]
Looks promising. I still think we will work on merge avoidance and other privacy techniques first that have lower implementation complexity and don't rely on a non-bitcoin network service to coordinate, but will keep an eye on this one.

**Comment by eragmus** [comment]
Fantastic :).

**Comment by Logicwax** [comment]
How about adding a manual override in settings to connect to one and only one node?     I run a full node, and various Android Bitcoin clients (like the "Bitcoin" one) allow me to choose to connect to my node only.    Letting me or trusted friends of mine have the ability for increased privacy by using a node we have control over.

**Comment by voisine** [comment]
@Logicwax, you can do that currently by building from source (it's a one line change), or by setting up your wi-fi router to redirect either port 8333 traffic or the dns seeds to your chosen node.

We're really focused on growing bitcoin adoption, and running your own full-node isn't the type of solution we think will become popular with the typical user, so that's why were exploring alternatives that will work for everyone.

**Comment by Logicwax** [comment]
That's an assumption that my node is at home on my network.   I'm talking about letting me connect to my node from anywhere I am.    This is available as an option in the few android wallets that exist.    

It's essentially the same use-case as electrum users.

**Comment by voisine** [comment]
Yes, in that situation you'd either have to set up a VPN on you phone, or change that line of code and build from source.

**Comment by Logicwax** [comment]
Yes, but why not add this as something buried in the settings?   This is a thread about privacy enhancing features.    All it takes is a textfield in settings for me to enter a hostname for a node to connect to.   A lot of people I've spoken to have said they would love that feature as they could all point their breadwallets to their own node for privacy reasons.

**Comment by e4xit** [comment]
Is this still ongoing? I would love to see Tor integration. I am currently having difficulty in connecting Bread to my own node as it always has all connection slots full (35 max). I can't whitelist my ip address as of course being a mobile wallet it is always changing... The way I have been recommended to make it work is to set my node to run as a Tor Hidden Service and connect that way, but it seems Bread (still) doesn't support that?
